File: WorldFeed/Server/src/WorldFeed.Controllers/Upload/ChunkedUploader.cs
namespace WorldFeed.Controllers.Upload
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;
    using Newtonsoft.Json;

    using WorldFeed.Common.DTO;
    using WorldFeed.Common.Public;
    using WorldFeed.Common.Public.Events;
    using WorldFeed.Common.Public.Models.Enums;
    using WorldFeed.Common.Public.Models.Interfaces;
    using WorldFeed.Common.Public.Models.Interfaces.DTO;
    using WorldFeed.Common.Public.Parameters;
    using WorldFeed.Common.Public.Parameters.Enums;
    using WorldFeed.Common.QueryGenerators;
    using WorldFeed.Common.Upload;
    using WorldFeed.Common.Web;

    public class ChunkedUploader : IChunkedUploader
    {
        private readonly IMedia media;
        private readonly ITwitterAccessor twitterAccessor;
        private readonly IUploadQueryGenerator uploadQueryGenerator;
        private readonly ChunkUploadResult result;

        private int? expectedBinaryLength;

        public ChunkedUploader(ITwitterAccessor twitterAccessor, IUploadQueryGenerator uploadQueryGenerator, IMedia media)
        {
            this.twitterAccessor = twitterAccessor;
            this.uploadQueryGenerator = uploadQueryGenerator;
            this.media = media;
            this.result = new ChunkUploadResult
            {
                Media = this.media
            };

            UploadedSegments = new Dictionary<long, byte[]>();
        }

        public long? MediaId
        {
            get => this.media.Id;
            set => this.media.Id = value;
        }

        public Dictionary<long, byte[]> UploadedSegments { get; }

        public int NextSegmentIndex { get; set; }

        public async Task<bool> InitAsync(IChunkUploadInitParameters initParameters, ITwitterRequest request)
        {
            var initQuery = this.uploadQueryGenerator.GetChunkedUploadInitQuery(initParameters);

            request.Query.Url = initQuery;
            request.Query.HttpMethod = HttpMethod.POST;

            var twitterResult = await this.twitterAccessor.ExecuteRequestAsync<IUploadInitModel>(request).ConfigureAwait(false);
            this.result.Init = twitterResult;

            var initModel = twitterResult?.Model;
            if (initModel != null)
            {
                this.expectedBinaryLength = initParameters.TotalBinaryLength;
                this.media.Id = initModel.MediaId;
            }

            return initModel != null;
        }

        public async Task<bool> AppendAsync(IChunkUploadAppendParameters parameters, ITwitterRequest request)
        {
            if (this.MediaId == null)
            {
                throw new InvalidOperationException("You cannot append content to a non initialized chunked upload. You need to invoke the initialize method OR set the MediaId property of an existing ChunkedUpload.");
            }

            if (parameters.SegmentIndex == null)
            {
                parameters.SegmentIndex = this.NextSegmentIndex;
            }

            if (parameters.MediaId == null)
            {
                parameters.MediaId = this.MediaId;
            }

            var appendQuery = this.uploadQueryGenerator.GetChunkedUploadAppendQuery(parameters);

            var multipartQuery = new MultipartTwitterQuery(request.Query)
            {
                Url = appendQuery,
                HttpMethod = HttpMethod.POST,
                Binaries = new[] { parameters.Binary },
                Timeout = parameters.Timeout ?? TimeSpan.FromMilliseconds(System.Threading.Timeout.Infinite),
                ContentId = parameters.MediaType,
                UploadProgressChanged = args =>
                {
                    var progressChangedEventArgs = new MediaUploadProgressChangedEventArgs(UploadProgressState.PROGRESS_CHANGED, args.NumberOfBytesUploaded, args.TotalOfBytesToUpload);
                    parameters.UploadProgressChanged(progressChangedEventArgs);
                },
            };

            request.Query = multipartQuery;

            var twitterResult = await this.twitterAccessor.ExecuteRequestAsync(request).ConfigureAwait(false);
            this.result.AppendsList.Add(twitterResult);

            if (twitterResult.Response.IsSuccessStatusCode)
            {
                this.UploadedSegments.Add(parameters.SegmentIndex.Value, parameters.Binary);
                ++NextSegmentIndex;
            }

            return twitterResult.Response.IsSuccessStatusCode;
        }

        public async Task<bool> FinalizeAsync(ICustomRequestParameters customRequestParameters, ITwitterRequest request)
        {
            if (this.MediaId == null)
            {
                throw new InvalidOperationException("You cannot complete a non initialized chunked upload. Please initialize the method, append some content and then complete the upload.");
            }

            var finalizeQuery = this.uploadQueryGenerator.GetChunkedUploadFinalizeQuery(MediaId.Value, customRequestParameters);

            request.Query.Url = finalizeQuery;
            request.Query.HttpMethod = HttpMethod.POST;

            var finalizeTwitterResult = await this.twitterAccessor.ExecuteRequestAsync<UploadedMediaInfo>(request).ConfigureAwait(false);
            var uploadedMediaInfos = finalizeTwitterResult.Model;

            this.UpdateMedia(uploadedMediaInfos);

            this.result.Finalize = finalizeTwitterResult;

            return finalizeTwitterResult.Response.IsSuccessStatusCode;
        }

        public IChunkUploadResult Result => this.result;

        private void UpdateMedia(IUploadedMediaInfo uploadedMediaInfos)
        {
            this.media.UploadedMediaInfo = uploadedMediaInfos;

            if (this.expectedBinaryLength != null)
            {
                // If all the data has not been sent then we do not construct the data
                if (UploadedSegments.Sum(x => x.Value.Length) == this.expectedBinaryLength)
                {
                    var allSegments = UploadedSegments.OrderBy(x => x.Key);
                    this.media.Data = allSegments.SelectMany(x => x.Value).ToArray();
                }
            }
        }

        // ReSharper disable once ClassNeverInstantiated.Local
        private class UploadInitModel
        {
            [JsonProperty("media_id")]
            public long MediaId { get; set; }


            [JsonProperty("expires_after_secs")]
            public long ExpiresAfterInSeconds { get; set; }
        }
    }
}

File: WorldFeed/Client/src/app/controllers/Upload/ChunkedUploader.ts
import {IChunkedUploader} from "../../core/Core/Upload/IChunkedUploader";
import {IMedia} from '../../core/Public/Models/Interfaces/IMedia';
import {IUploadQueryGenerator} from '../../core/Core/QueryGenerators/IUploadQueryGenerator';
import {ChunkUploadResult, IChunkUploadResult} from "../../core/Core/Upload/ChunkUploaderResult";
import Dictionary from "../../c#-objects/TypeScript.NET-Core/packages/Core/source/Collections/Dictionaries/Dictionary";
import {IChunkUploadInitParameters} from "../../core/Core/Upload/ChunkUploadInitParameters";
import {ITwitterRequest} from '../../core/Public/Models/Interfaces/ITwitterRequest';
import {HttpMethod} from "../../core/Public/Models/Enum/HttpMethod";
import {IUploadInitModel} from "../../core/Core/Upload/UploadInitModel";
import InvalidOperationException from 'src/app/c#-objects/TypeScript.NET-Core/packages/Core/source/Exceptions/InvalidOperationException';
import {IChunkUploadAppendParameters} from "../../core/Core/Upload/ChunkUploadAppendParameters";
import {MultipartTwitterQuery} from "../../core/Public/MultipartTwitterQuery";
import {ICustomRequestParameters} from "../../core/Public/Parameters/CustomRequestParameters";
import {IUploadedMediaInfo} from "../../core/Public/Models/Interfaces/DTO/IUploadedMediaInfo";
import TimeSpan from "../../c#-objects/TypeScript.NET-Core/packages/Core/source/Time/TimeSpan";
import {UploadProgressState} from "../../core/Public/Parameters/Enum/UploadProgressState";
import {ITwitterAccessor} from "../../core/Core/Web/ITwitterAccessor";
import {MediaUploadProgressChangedEventArgs} from "../../core/Public/Events/MediaUploadProgressChangedEventArgs";
import {ITwitterResult} from "../../core/Core/Web/TwitterResult";
import {UploadedMediaInfo} from "../../core/Core/DTO/UploadedMediaInfo";

export class ChunkedUploader implements IChunkedUploader {
  private readonly _media: IMedia;
  private readonly _twitterAccessor: ITwitterAccessor;
  private readonly _uploadQueryGenerator: IUploadQueryGenerator;

  private _expectedBinaryLength?: number;
  private readonly _result: ChunkUploadResult;

  constructor(twitterAccessor: ITwitterAccessor, uploadQueryGenerator: IUploadQueryGenerator, media: IMedia) {
    this._twitterAccessor = twitterAccessor;
    this._uploadQueryGenerator = uploadQueryGenerator;
    this._media = media;

    this._result = new ChunkUploadResult();
    this._result.media = this._media;

    this.uploadedSegments = new Dictionary<number, number[]>();
  }

  get mediaId(): number {
    return this._media.id;
  }

  set mediaId(value: number) {
    this._media.id = value;
  }

  public uploadedSegments: Dictionary<number, number[]>;
  public nextSegmentIndex: number;

  public async initAsync(initParameters: IChunkUploadInitParameters, request: ITwitterRequest): Promise<boolean> {
    let initQuery = this._uploadQueryGenerator.getChunkedUploadInitQuery(initParameters);

    request.query.url = initQuery;
    request.query.httpMethod = HttpMethod.POST;

    let twitterResult = await this._twitterAccessor.executeRequestAsync<IUploadInitModel>(request); // .ConfigureAwait(false);
    this._result.init = twitterResult;

    let initModel: IUploadInitModel = twitterResult?.model;
    if (initModel != null) {
      this._expectedBinaryLength = initParameters.totalBinaryLength;
      this._media.id = initModel.mediaId;
    }

    return initModel != null;
  }

  public async appendAsync(parameters: IChunkUploadAppendParameters, request: ITwitterRequest): Promise<boolean> {
    if (this.mediaId == null) {
      throw new InvalidOperationException(`You cannot append content to a non initialized chunked upload.
                 You need to invoke the initialize method OR set the MediaId property of an existing ChunkedUpload.`);
    }

    if (parameters.segmentIndex == null) {
      parameters.segmentIndex = this.nextSegmentIndex;
    }

    if (parameters.mediaId == null) {
      parameters.mediaId = this.mediaId;
    }

    let appendQuery: string = this._uploadQueryGenerator.getChunkedUploadAppendQuery(parameters);

    let multipartQuery = new MultipartTwitterQuery(request.query);
    multipartQuery.url = appendQuery;
    multipartQuery.httpMethod = HttpMethod.POST;
    multipartQuery.Binaries = [parameters.binary];              // new[] { parameters.Binary };
    multipartQuery.timeout = parameters.timeout ?? TimeSpan.fromMilliseconds(-1); // Timeout.Infinite
    multipartQuery.ContentId = parameters.mediaType;
    multipartQuery.UploadProgressChanged = args => {
      let progressChangedEventArgs = new MediaUploadProgressChangedEventArgs(UploadProgressState.PROGRESS_CHANGED, args.NumberOfBytesUploaded, args.TotalOfBytesToUpload);
      parameters.uploadProgressChanged(progressChangedEventArgs);
    };

    request.query = multipartQuery;

    let twitterResult: ITwitterResult = await this._twitterAccessor.executeRequestAsync(request); // .ConfigureAwait(false);
    this._result.appendsList.push(twitterResult);

    if (twitterResult.response.isSuccessStatusCode) {
      this.uploadedSegments.setValue(parameters.segmentIndex/*.Value*/, parameters.binary);
      ++this.nextSegmentIndex;
    }

    return twitterResult.response.isSuccessStatusCode;
  }

  public async finalizeAsync(customRequestParameters: ICustomRequestParameters, request: ITwitterRequest): Promise<boolean> {
    if (this.mediaId == null) {
      throw new InvalidOperationException(`You cannot complete a non initialized chunked upload. Please
                                                             initialize the method, append some content and then complete the upload.`);
    }

    let finalizeQuery = this._uploadQueryGenerator.getChunkedUploadFinalizeQuery(this.mediaId/*.Value*/, customRequestParameters);

    request.query.url = finalizeQuery;
    request.query.httpMethod = HttpMethod.POST;

    let finalizeTwitterResult = await this._twitterAccessor.executeRequestAsync<UploadedMediaInfo>(request); // .ConfigureAwait(false);
    let uploadedMediaInfos = finalizeTwitterResult.model;

    this.updateMedia(uploadedMediaInfos);

    this._result.finalize = finalizeTwitterResult;

    return finalizeTwitterResult.response.isSuccessStatusCode;
  }


  get result(): IChunkUploadResult {
    return this._result;
  }

  private updateMedia(uploadedMediaInfos: IUploadedMediaInfo): void {
    this._media.uploadedMediaInfo = uploadedMediaInfos;

    if (this._expectedBinaryLength != null) {
      let binaryLengthCurrent;
      for (let i = 0; i < this.uploadedSegments.count; i++) {
        binaryLengthCurrent += this.uploadedSegments.getValue(i).length;   // TODO: might bug!!
      }
      // If all the data has not been sent then we do not construct the data
      if (binaryLengthCurrent === this._expectedBinaryLength) {
        let allSegments = this.uploadedSegments.toArray().sort((a, b) => a.key - b.key); // TODO: bug debug it
        this._media.data = allSegments.reduce((a, b) => [...a, ...b.value], []); // .ToArray();
      }
    }
  }

  private UploadInitModel = class {
    // [JsonProperty("media_id")];
    public MediaId: number;

    // [JsonProperty("expires_after_secs")];
    public ExpiresAfterInSeconds: number;
  };
}


